File: Data Structures & Algorithms/rotting-fruit/submission-2.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        q = deque()
        rows = len(grid)
        cols = len(grid[0])
        visited = set()
        fresh = 0

        # Add fresh fruit to the queue.
        def add_queue(r, c):
            nonlocal fresh
            if (r < 0 or r >= rows or
                c < 0 or c >= cols or
                (r, c) in visited or
                grid[r][c] == 2 or
                grid[r][c] == 0):
                return

            fresh -= 1
            q.append((r, c))
            visited.add((r, c))

        # Store all rotten fruits in a queue
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 2:
                    q.append((r, c))
                    visited.add((r, c))
                elif grid[r][c] == 1:
                    fresh += 1

        # Traverse layer by layer
        # Add the fresh fruit to the q every minute.
        mins = 0
        while fresh > 0 and q:
            for i in range(len(q)):
                r, c = q.popleft()
                grid[r][c] = 2
                add_queue(r + 1, c)
                add_queue(r - 1, c)
                add_queue(r, c + 1)
                add_queue(r, c - 1)

            mins += 1
        
        # Check if there is any fresh fruit.
        if fresh == 0:
            return mins
        return -1

File: Data Structures & Algorithms/surrounded-regions/submission-2.cpp
class Solution {
private:
    int rows, cols;

    void capture(int r, int c, vector<vector<char>>& board) {
        if (r < 0 ||
            r >= rows ||
            c < 0 ||
            c >= cols ||
            board[r][c] != 'O') {
                return;
        }

        board[r][c] = 'T';

        capture(r + 1, c, board);
        capture(r - 1, c, board);
        capture(r, c + 1, board);
        capture(r, c - 1, board);
    }

public:
    void solve(vector<vector<char>>& board) {
        rows = board.size();
        cols = board[0].size();

        for (int r = 0; r < rows; r++) {
            if (board[r][0] == 'O') {
                capture(r, 0, board);
            }
        }

        for (int c = 0; c < cols; c++) {
            if (board[0][c] == 'O') {
                capture(0, c, board);
            }
        }

        for (int r = 0; r < rows; r++) {
            if (board[r][cols - 1] == 'O') {
                capture(r, cols - 1, board);
            }
        }

        for (int c = 0; c < cols; c++) {
            if (board[rows - 1][c] == 'O') {
                capture(rows - 1, c, board);
            }
        }

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (board[r][c] == 'O') {
                    board[r][c] = 'X';
                } else if (board[r][c] == 'T') {
                    board[r][c] = 'O';
                }
            }
        }
    }
};
